### ISSUE TITLE
Problem: Bad gateway 502 AM 1.9

### ISSUE THREAD

**Issue by bcitamatica** [archivematica/Issues#600]
**Expected behaviour**
Dasboard service started

**Current behaviour**
Dashboard service does not start automatically. Needs manual restart

**Steps to reproduce**
Reboot

**Your environment (version of Archivematica, OS version, etc)**
AM1.9 Ubuntu Server 18.04 LTS

Got this error twice in a row....Installation works fine and I can connect fine to both dashboard and storage service...

But after reboot on Ubuntu 18 I get 502 bad gateway on the dashboard. I can still connect fine to port 8000 (storage)

Error log gives me :

2019/03/12 13:27:36 [error] 1359#1359: *24 connect() failed (111: Connection refused) while connecting to upstream, client: 142.232.42.26, server: _, request: "GET / HTTP/1.1", upstream: "http://127.0.0.1:8002/", host: "142.232.42.25"

---
**For Artefactual use:**
Please make sure these steps are taken before moving this issue from Review to Verified in Waffle:

- All PRs related to this issue are properly linked 👍
- All PRs related to this issue have been merged 👍
- Test plan for this issue has been implemented and passed 👍
- Documentation regarding this issue has been written and it has been added to the release notes, if needed 👍

**Comment by ThomasEdvardsen** [comment]
The same is happening on CentOS 7.6.1810 as well. Workaround is manually starting dashbord service after reboot.

**Comment by goetzk** [comment]
I'm not sure if this problem is planned for #581 but its coming up a lot in the group (https://groups.google.com/forum/#!forum/archivematica) - users are faced with an  Internal Server Error / 502 bad gateway on a regular basis; some of the threads below include multiple people with the same issue.

This has been mentioned in numerous threads; they include more detail:
https://groups.google.com/forum/#!topic/archivematica/GmNN1DP9uoQ
https://groups.google.com/forum/#!topic/archivematica/6415sbzZbpE
https://groups.google.com/forum/#!topic/archivematica/uXg7x__pgSs

I'm not sure if this one is related or not
https://groups.google.com/forum/#!topic/archivematica/GMwE89zyhj8

From the outside it looks to me like a timing issue, but I haven't installed 1.9 to confirm that.

Karl.

**Comment by mamedin** [comment]
I could reproduce the issue restarting an  AM1.9 server. The dashboard has not started because the elasticsearch server was up but not listening in port 9200. 

Dashboard log:
<pre>
artefactual@am191bionic:~$ sudo tail -n30 /var/log/archivematica/dashboard/dashboard.log 
  File "/usr/share/archivematica/virtualenvs/archivematica-dashboard/local/lib/python2.7/site-packages/urllib3/connection.py", line 168, in _new_conn
    self, "Failed to establish a new connection: %s" % e)
NewConnectionError: <urllib3.connection.HTTPConnection object at 0x7fbe643cbad0>: Failed to establish a new connection: [Errno 111] Connection refused
WARNING   2019-03-27 10:05:26  elasticsearch:base:log_request_fail:97:  HEAD http://127.0.0.1:9200/aips,aipfiles,transfers,transferfiles [status:N/A request:0.001s]
Traceback (most recent call last):
  File "/usr/share/archivematica/virtualenvs/archivematica-dashboard/local/lib/python2.7/site-packages/elasticsearch/connection/http_urllib3.py", line 172, in perform_request
    response = self.pool.urlopen(method, url, body, retries=Retry(False), headers=request_headers, **kw)
  File "/usr/share/archivematica/virtualenvs/archivematica-dashboard/local/lib/python2.7/site-packages/urllib3/connectionpool.py", line 638, in urlopen
    _stacktrace=sys.exc_info()[2])
  File "/usr/share/archivematica/virtualenvs/archivematica-dashboard/local/lib/python2.7/site-packages/urllib3/util/retry.py", line 343, in increment
    raise six.reraise(type(error), error, _stacktrace)
  File "/usr/share/archivematica/virtualenvs/archivematica-dashboard/local/lib/python2.7/site-packages/urllib3/connectionpool.py", line 600, in urlopen
    chunked=chunked)
  File "/usr/share/archivematica/virtualenvs/archivematica-dashboard/local/lib/python2.7/site-packages/urllib3/connectionpool.py", line 354, in _make_request
    conn.request(method, url, **httplib_request_kw)
  File "/usr/lib/python2.7/httplib.py", line 1042, in request
    self._send_request(method, url, body, headers)
  File "/usr/lib/python2.7/httplib.py", line 1082, in _send_request
    self.endheaders(body)
  File "/usr/lib/python2.7/httplib.py", line 1038, in endheaders
    self._send_output(message_body)
  File "/usr/lib/python2.7/httplib.py", line 882, in _send_output
    self.send(msg)
  File "/usr/lib/python2.7/httplib.py", line 844, in send
    self.connect()
  File "/usr/share/archivematica/virtualenvs/archivematica-dashboard/local/lib/python2.7/site-packages/urllib3/connection.py", line 181, in connect
    conn = self._new_conn()
  File "/usr/share/archivematica/virtualenvs/archivematica-dashboard/local/lib/python2.7/site-packages/urllib3/connection.py", line 168, in _new_conn
    self, "Failed to establish a new connection: %s" % e)
NewConnectionError: <urllib3.connection.HTTPConnection object at 0x7fbe643cbb10>: Failed to establish a new connection: [Errno 111] Connection refused
</pre>

**Comment by mamedin** [comment]
A possible solution is to add the following lines to the corresponding sections at `/etc/systemd/system/archivematica-dashboard.service` file:

For [Unit] section:

<pre>
StartLimitInterval=200
StartLimitBurst=5
</pre>

For [Service] section:

<pre>
Restart=always
RestartSec=30
</pre>

And reload systemd daemon:

<pre>
sudo systemctl daemon-reload
</pre>

This way the systemd daemon allows 5 retries separated by 30 seconds.

**Comment by scollazo** [comment]
@mamedin I like that approach for ansible/package deployments, but we should also have in mind docker deployments, where we can't rely on systemd

**Comment by goetzk** [comment]
@mamedin My suggestion was to use a dependency rather then waiting but that also assumes systemd.

@scollazo If we can't rely on the init system it sounds like all the archivematica services will need to re implement some form of socket based activation or wait-until-available. I see from a search that docker-compose can start services in dependency order so if adding retry code to all daemons isn't going to happen the fix may need to be made systemd and docker.

**Comment by mamedin** [comment]
@goetzk the dependency with "Wants" and "After" has a problem with elasticsearch service because it returns to systemd it is UP but the port 9200 is not listening yet.

**Comment by bcitamatica** [comment]
I do not have an /etc/systemd/system/archivematica-dashboard.service  file ?!
Am 1.9 Ubuntu 18.04.

**Comment by mamedin** [comment]
@bcitamatica and `/etc/systemd/system/multi-user.target.wants/archivematica-dashboard.service` file?

**Comment by EagleRock1313** [comment]
I am working on a manual install today. UbuntuServer 18.0.4 + ES 7.1.0 + AM 1.9.1.  
The below changes seemed to fix the error. This is my first time installing AM.

- update file archivematica-dashboard.service per google groups thread here
https://groups.google.com/forum/#!searchin/archivematica/502$20bad$20gateway%7Csort:date/archivematica/6415sbzZbpE/s4Qm5d4sAAAJ

- Needed to edit ES config file at /etc/elasticsearch/elasticsearch.yml
uncomment and add ip address > network.host <ip address>
uncomment > http.port: 9200
- Reference: https://www.elastic.co/guide/en/elasticsearch/reference/current/deb.html

- also added these to crontab -e with @reboot <command>
@reboot sudo systemctl daemon-reload
@reboot sudo service elasticsearch restart
@reboot sudo systemctl enable elasticsearch

@reboot sudo service clamav-freshclam restart
@reboot sleep 120s
@reboot sudo service clamav-daemon start
@reboot sudo service gearman-job-server restart
@reboot sudo service archivematica-mcp-server start
@reboot sudo service archivematica-mcp-client restart
@reboot sudo service archivematica-storage-service start
@reboot sudo service archivematica-dashboard restart
@reboot sudo service nginx restart
@reboot sudo systemctl enable fits-nailgun
@reboot sudo service fits-nailgun start

Not sure what tomorrow will bring. I will be motoring around the dashboard to see if I broke anything.

**Comment by mamedin** [comment]
Hi @EagleRock1313 , thanks for your comment :)

Please use ElasticSearch 6 (we usually install ES 6.5.4 in artefactual deployments). ElasticSearch 7 has not been tested with Archivematica, so you could find some issues.

**Comment by joel-simpson** [comment]
@mamedin I see you merged two PRs to address this defect... if there is no more work to be done can you please move this over to status "Review"?

**Comment by EagleRock1313** [comment]
Thank you Joel/Miguel. I will downgrade my ElasticSearch and try reconfiguring.

**Comment by mamedin** [comment]
> @mamedin I see you merged two PRs to address this defect... if there is no more work to be done can you please move this over to status "Review"?

Done, but I don't know if docker deploys need a change.

**Comment by dbqreference** [comment]
I'm stuck with this problem after a fresh install and not sure what else to do. I've implemented the workaround changes in the configuration files, restarted and reconfigured each of the pieces mentioned in the GoogleGroup and mentioned above. 

What other options do I have moving forward?

**Comment by EagleRock1313** [comment]
Hi dbqreference,
I am stuck too, Bad Gateway continues.  I am going to downgrade to Elasticsearch V6 from V7.  Per mamedin above, V7 is untested with AM 1.9.  As of today, I have not had time to do the ES downgrade to V6.

**Comment by mamedin** [comment]
@dbqreference and @EagleRock1313 after restarting (manual mode) the dashboard service, is the dashboard up and running or your problem is that dashboard never starts?

**Comment by dbqreference** [comment]
@mamedin dashboard never starts, I'm stuck at the 502 bad gateway. I've already configured my security groups on AWS to allow traffic from the public IP we have for it and I double checked to make sure I'm running Elasticsearch V6 so I didn't have to worry about the downgrade.

**Comment by mamedin** [comment]
@dbqreference I think you should check:

* Elasticsearch can be used from dashboard server. For example running from AM server: 
```
curl -XGET 'http://ES_SERVER_IP_ADDRESS:9200/_cluster/health?pretty=true'
```
* Dashboard logs and `systemctl status archivematica-dashboard`. 

The above workarounds fixed the issue: elasticsearch tried to start before elasticsearch service was up and running. I suspect you have other issue, but the same 502 error (dashboard service down).

**Comment by dbqreference** [comment]
@mamedin I'm getting the following error;
`curl: (7) Failed to connect to SERVER_IP:9200; Connection refused`
I'm certain I changed all the necessary things for this port based on what @EagleRock1313 said about uncommenting the yml file for the http_port:9200

**Comment by dbqreference** [comment]
Here's what the status fed back:
`archivematica-dashboard.service - Archivematica Dashboard
   Loaded: loaded (/usr/lib/systemd/system/archivematica-dashboard.service; enabled; vendor preset: disabled)
   Active: activating (auto-restart) (Result: exit-code) since Thu 2019-06-13 15:53:28 UTC; 9s ago
  Process: 28758 ExecStop=/bin/kill -s TERM $MAINPID (code=exited, status=1/FAILURE)
  Process: 28743 ExecStart=/usr/share/archivematica/virtualenvs/archivematica-dashboard/bin/gunicorn --config /etc/archivematica/dashboard.gunicorn-config.py wsgi:application (code=exited, status=3)
 Main PID: 28743 (code=exited, status=3)

Jun 13 15:53:28 IPADDRESS: archivematica-dashboard.service: control process exited, code=exited status=1
Jun 13 15:53:28 IPADDRESS: Unit archivematica-dashboard.service entered failed state.
Jun 13 15:53:28 IPADDRESS: archivematica-dashboard.service failed.
`

**Comment by mamedin** [comment]
@dbqreference so you need to review the ES and network settings to allow the ElastiSearch connections from Archivematica server.

**Comment by dbqreference** [comment]
@mamedin If I may be so bold as to inquire how I would go about doing that, I would appreciate any help in this matter.

**Comment by mamedin** [comment]
@dbqreference I will help you. Please, can you open a new thread in GoogleGroup about ElasticSearch settings when this service is running in a separate server? I think it can help to other AM users.

**Comment by dbqreference** [comment]
@mamedin here's the link: https://groups.google.com/d/msg/archivematica/UQ50s5Onybw/UwcT1UTEDAAJ

**Comment by ablwr** [comment]
This should be resolved in 1.10 onwards.
